bee-api/bee-rest-api/src/types/dtos.rs

```diff
@@ -554,6 +707,476 @@ pub struct ReferenceUnlockBlockDto {
+        value
+            .0
+            .parse::<TokenId>()
+            .map_err(|_| Error::InvalidSemanticField("token id"))
+    }
+}
+
+/// Describes an U256.
+#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Serialize, Deserialize)]
+pub struct U256Dto(pub String);
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum FeatureBlockDto {
+    /// A sender feature block.
+    Sender(SenderFeatureBlockDto),
+    /// An issuer feature block.
+    Issuer(IssuerFeatureBlockDto),
+    /// A dust deposit return feature block.
+    DustDepositReturn(DustDepositReturnFeatureBlockDto),
+    /// A timelock milestone index feature block.
+    TimelockMilestoneIndex(TimelockMilestoneIndexFeatureBlockDto),
+    /// A timelock unix feature block.
+    TimelockUnix(TimelockUnixFeatureBlockDto),
+    /// An expiration milestone index feature block.
+    ExpirationMilestoneIndex(ExpirationMilestoneIndexFeatureBlockDto),
+    /// An expiration unix feature block.
+    ExpirationUnix(ExpirationUnixFeatureBlockDto),
+    /// An indexation feature block.
+    Indexation(IndexationFeatureBlockDto),
+    /// A metadata feature block.
+    Metadata(MetadataFeatureBlockDto),
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SenderFeatureBlockDto(pub AddressDto);
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct IssuerFeatureBlockDto(pub AddressDto);
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct DustDepositReturnFeatureBlockDto(pub u64);
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct TimelockMilestoneIndexFeatureBlockDto(pub MilestoneIndex);
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct TimelockUnixFeatureBlockDto(pub u32);
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct ExpirationMilestoneIndexFeatureBlockDto(pub MilestoneIndex);
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct ExpirationUnixFeatureBlockDto(pub u32);
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct IndexationFeatureBlockDto(pub String);
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct MetadataFeatureBlockDto(pub String);
+
+impl FeatureBlockDto {
+    /// Return the output kind of an `Output`.
+    pub fn kind(&self) -> u8 {
+        match self {
+            Self::Sender(_) => SenderFeatureBlock::KIND,
+            Self::Issuer(_) => IssuerFeatureBlock::KIND,
+            Self::DustDepositReturn(_) => DustDepositReturnFeatureBlock::KIND,
+            Self::TimelockMilestoneIndex(_) => TimelockMilestoneIndexFeatureBlock::KIND,
+            Self::TimelockUnix(_) => TimelockUnixFeatureBlock::KIND,
+            Self::ExpirationMilestoneIndex(_) => ExpirationMilestoneIndexFeatureBlock::KIND,
+            Self::ExpirationUnix(_) => ExpirationUnixFeatureBlock::KIND,
+            Self::Indexation(_) => IndexationFeatureBlock::KIND,
+            Self::Metadata(_) => MetadataFeatureBlock::KIND,
+        }
+    }
+}
+
+impl From<&FeatureBlock> for FeatureBlockDto {
```

Comment:
I think `From<T>` is more flexible than `From<&T>`:

- If we implement `From<T>` those two are fine:

```rust
let t: T = ...;
let v = V::from(t); // We don't clone `t` because we won't use it anymore
```

```rust
let t: T = ...;
let v = V::from(t.clone()); // We clone `t` because we want to use it later
t.do_stuff();
```

- If we implement `From<&T>` then:

```rust
let t: T = ...;
let v = V::from(&t); // This will clone/copy stuff from `t` even though `t` won't be used anymore.
```

```rust
let t: T = ...;
let v = V::from(&t);
t.do_stuff();
```

Reply:
Do you generally think we should do `From<T>` instead of  `From<&T>` or just for the `FeatureBlock` here?

Not sure I understand your reasoning:
```
let t: T = ...;
let v = V::from(&t); // This will clone/copy stuff from `t` even though `t` won't be used anymore.
```
Why don't I want to use `t` anymore? 

Whats your take on this @thibault-martinez ?

Reply:
I'd do everywhere. If we need `From<&T>` we should be implementing `AsRef<T>`, `Borrow<T>` or `Deref<T>`.